Add multi-segment capability to BaseRasterWidget and children

jakeswann1 opened the pull request:
Adds the option to pass a list of segment indices to the `AmplitudesWidget`, `DriftRasterMapWidget`, and `RasterWidget`  to plot across multiple segments, by updating how the base widget handles plotting data. Maintains current default behaviour and SortingView capability. resolves #3801

jakeswann1 commented:
resolves #3801

jakeswann1 commented:
Multi-segment plots would look like this:

![image](https://github.com/user-attachments/assets/e8e1d861-cdfd-4fdb-be52-9565595e4cb7)
![image](https://github.com/user-attachments/assets/cf44cb9f-3db4-4988-bff6-95e241140946)
![image](https://github.com/user-attachments/assets/a11e5e9e-0924-4059-b4b4-7cc6545b9626)

chrishalcrow commented on src/spikeinterface/widgets/rasters.py:
```diff
@@ -15,12 +15,17 @@ class BaseRasterWidget(BaseWidget):
+        converted to a dict of dicts with segment 0.
     unit_ids : array-like | None, default: None
         List of unit_ids to plot
+    segment_index : int | list | None, default: None
```
I would simplify this by using `segment_indices` and only allow `list | None`. Two advantages: 1) as a user it's not obvious that `segment_index` can take lists since it's singular. 2) it simplifies the code for checking that the segments exist lower down.

chrishalcrow commented on src/spikeinterface/widgets/rasters.py:
```diff
@@ -71,13 +79,105 @@ def __init__(
+            # Multiple segments specified
+            for idx in segment_index:
+                if idx not in available_segments:
+                    raise ValueError(f"segment_index {idx} not found in data")
```
Would be good to display the available segments. Something like
```
raise ValueError(f"segment_index {idx} not found in available segments {available_segments}")
```

chrishalcrow commented on src/spikeinterface/widgets/amplitudes.py:
```diff
@@ -64,53 +65,96 @@ def __init__(
+        # Calculate total duration across all segments for x-axis limits
+        total_duration = 0
+        for idx in segment_indices:
+            duration = sorting_analyzer.get_num_samples(idx) / sorting_analyzer.sampling_frequency
```
I'd suggest we make a `durations` list here, and then update `BaseRasterWidget` to take a list of durations.

chrishalcrow commented on src/spikeinterface/widgets/rasters.py:
```diff
@@ -15,12 +15,17 @@ class BaseRasterWidget(BaseWidget):
+    segment_index : int | list | None, default: None
+        For multi-segment data, specifies which segment(s) to plot. If None, uses all available segments.
+        For single-segment data, this parameter is ignored.
     total_duration : int | None, default: None
```
Suggest replace `total_duration` with `durations` which takes a list of durations per segment

chrishalcrow commented on src/spikeinterface/widgets/rasters.py:
```diff
@@ -71,13 +79,105 @@ def __init__(
+            cumulative_durations.append(cumulative_durations[-1] + duration)
+
+        # Segment boundaries for visualization (only internal boundaries)
+        segment_boundaries = cumulative_durations[1:] if len(segments_to_use) > 1 else None
```
There's a numpy funciton for this. Something like
```
segment_boundaries = np.cumsum(segment_durations)
cumulative_durations = np.concatenate([[0], segment_boundaries])
```
should do the job!

chrishalcrow commented on src/spikeinterface/widgets/rasters.py:
```diff
@@ -71,13 +79,105 @@ def __init__(
+                    spike_times = spike_train_data[seg_idx][unit_id]
+
+                    # Adjust spike times by adding cumulative duration of previous segments
+                    if offset > 0:
```
Is the only case where `offset <= 0` when it is equal to `0`? If so, it's ok to do `adjusted_times = spike_times + 0`. Then we can replace these four lines with just 
```
adjusted_times = spike_times + offset
```

chrishalcrow commented on src/spikeinterface/widgets/rasters.py:
```diff
@@ -71,13 +79,105 @@ def __init__(
+                    y_values = y_axis_data[seg_idx][unit_id]
+
+                    # Concatenate with any existing data
+                    if len(concatenated_spike_trains[unit_id]) > 0:
```
it's ok to concatenate a full array to an empty list. So I think the stuff under `else` is redundant, and you can remove the `if` statement

chrishalcrow commented on src/spikeinterface/widgets/rasters.py:
```diff
@@ -71,13 +79,105 @@ def __init__(
+        concatenated_spike_trains = {unit_id: [] for unit_id in unit_ids}
+        concatenated_y_axis = {unit_id: [] for unit_id in unit_ids}
+
+        for i, seg_idx in enumerate(segments_to_use):
```
You can simplify some of the gross indexing in the for loop by using some `zip`ping here. Try a `for` loop that looks like:

``` python
for offset, spike_train_segment, y_axis_segment in zip(cumulative_durations, spike_train_data.values(), y_axis_data.values()):
    for unit_id, spike_times in spike_train_segment.items():
        y_values = y_axis_segment[unit_id]
```
Should make the code more readable!

jakeswann1 commented:
Hey @chrishalcrow - sorry for the delay working on this, I've been away for a while. Think I've addressed your comments, let me know if there's anything else you think could be improved here!

chrishalcrow commented on src/spikeinterface/widgets/rasters.py:
```diff
@@ -296,46 +373,72 @@ def __init__(
+                    st = sorting.get_unit_spike_train(unit_id, segment_index=seg_idx, return_times=True)
+                    if len(st) > 0:
+                        max_time = max(max_time, np.max(st))
+                duration = max_time
```
It's a bit simpler to do this using the full spike train all at once. You get this using 
```
spikes = sorting.to_spike_vector()
```
then spikes has a `segment_index` which can be used to figure out the segment lengths:
```
segment_slices = [np.searchsorted(spikes["segment_index"], [segment_index, segment_index + 1]) for segment_index in segment_indices]
segment_lengths = [segment_slice[1] - segment_slice[0] for segment_slice in segment_slices
```
then you gotta convert to time using `sorting.sampling_frequency`.

Although it's a bit convoluted, I think using this would simplify the code a bit, and I'd be tempted to use it instead of the recording, because then we remove any mention of `recording`s in this function.

(This is a bit messy because we don't keep track of the segment length in the sorting object )

jakeswann1 replied:
I've made the change, but do you not think that removing the option to get segment duration from the recording object introduces inaccuracy because final spike time != recording end time? Obviously in most dense recordings this is a pretty minimal difference, but I'm not sure what the policy is more generally that you guys have been using to handle this

chrishalcrow commented on src/spikeinterface/widgets/motion.py:
```diff
@@ -226,18 +258,32 @@ def __init__(
+            if recording is not None and hasattr(recording, "get_duration"):
+                duration = recording.get_duration(seg_idx)
+            else:
+                # Estimate from spike times
```
Maybe check if you can use `np.searchsorted` to shorten this code, check the comments in `rasters.py` first.

chrishalcrow commented on src/spikeinterface/widgets/amplitudes.py:
```diff
@@ -64,57 +65,78 @@ def __init__(
+        # Calculate durations for all segments for x-axis limits
+        durations = []
+        for idx in segment_indices:
+            duration = sorting_analyzer.get_num_samples(idx) / sorting_analyzer.sampling_frequency
```
This will fail if `sorting_analyzer` doesn't have a `recording`. So we need to use a strategy that only uses the sorting, like the other functions.

chrishalcrow reviewed (changes requested):
Hey Jake! Looking amazing, and feels much simpler than before - thanks for all the updates.

I think all my feeback is about computing durations of segments. Have a read of the comments in `rasters.py` first, then the other stuff. Once that's sorted, this should be good to go!

chrishalcrow commented on src/spikeinterface/widgets/utils.py:
```diff
@@ -349,3 +351,48 @@ def make_units_table_from_analyzer(
     return units_table
+
+
+def validate_segment_indices(segment_indices: list[int] | None, sorting: BaseSorting):
```
Would be great to add a couple of little tests for this.

chrishalcrow commented on src/spikeinterface/widgets/amplitudes.py:
```diff
@@ -51,70 +52,104 @@ def __init__(
             bins = 100
 
+        # Calculate durations for all segments for x-axis limits
+        durations = get_segment_durations(sorting)
```
once function is updated, pass `segment_indices` here

chrishalcrow commented on src/spikeinterface/widgets/rasters.py:
```diff
@@ -320,30 +413,43 @@ def __init__(
+        spikes = sorting.to_spike_vector()
+
+        # Estimate segment duration from max spike time in each segment
+        durations = get_segment_durations(sorting)
```
once function is updated, pass `segment_indices` here

chrishalcrow commented:
One tiny change, then will approve :)

jakeswann1 commented:
Moved to #4035 so it can be merged properly
